File: src/antibody.cpp
#include <cpp11.hpp>
using namespace cpp11;

[[cpp11::register]]
doubles ab_cpp(integers t, doubles dose_timesteps, doubles init_titres, doubles prop_short, doubles dur_short, doubles dur_long) {
  int n = t.size();

  int n_set = dur_short.size();
  writable::doubles rs(n_set);
  writable::doubles rl(n_set);
  for(int i = 0; i < n_set; ++i) {
    rs[i] = log(2) / dur_short[i];
    rl[i] = log(2) / dur_long[i];
  }

  writable::doubles ab(n);
  int t_cur = 0;
  double init_titres_cur = init_titres[0];

  int t_index = 0;
  for(int i = 0; i < n; ++i) {  // Loop should start at 0 for proper indexing
    // Check if we need to increment t_index
    if (t_index + 1 < dose_timesteps.size() && t[i] == dose_timesteps[t_index + 1]) {
      t_index++;
    }

    // Check for down shifts on boostt
    if(t[i] > 1){
      if (t[i] == dose_timesteps[t_index]) {
        init_titres_cur = init_titres[t_index];
        if (ab[i - 1] > init_titres_cur) {
          init_titres_cur = ab[i - 1];
        }
      }
    }

    // Set t_cur relative to the current time division
    t_cur = t[i] - 1;
    if (t_index > 0) {
      t_cur = t[i] - dose_timesteps[t_index];
    }

    // Compute ab[i]
    ab[i] = init_titres_cur * (prop_short[t_index] * exp(-rs[t_index] * t_cur) + (1 - prop_short[t_index]) * exp(-rl[t_index] * t_cur));
  }

  return ab;
}

File: R/antibody.R
#' Estimate antibody titres
#'
#' @param timesteps Number of timesteps to output titres for
#' @param dose_timesteps Timesteps of doses
#' @param init_titres Titres at initital does or subsequent boost times
#' @param prop_short Proportion of the response that is short lived for each dose period
#' @param dur_short Duration of the short-lived response for each dose period
#' @param dur_long Duration of the long-lived response for each dose period
#' @param cpp Use cpp titre model
#'
#' @return Vector of antibody tires for 1:timesteps
#' @export
ab <- function(timesteps, dose_timesteps, init_titres, prop_short, dur_short, dur_long, cpp = TRUE){
  if(timesteps < 1){
    stop("timesteps must be positive")
  }
  if(!(all(dose_timesteps > 0) && all(diff(dose_timesteps) >= 0))){
    stop("dose_timesteps must be positive and montonically increasing")
  }
  if(any(init_titres <= 0)){
    stop("All init_titres must be > 0")
  }
  if(any(prop_short < 0) | any(prop_short > 1)){
    stop("prop_short must be between 0 and 1")
  }
  if(any(dur_short > dur_long)){
    stop("dur_short should be < dur_long")
  }

  timesteps <- 1:timesteps

  if(cpp){
    titres <- ab_cpp(timesteps, dose_timesteps, init_titres, prop_short, dur_short, dur_long)
  } else {
    titres <- ab_r(timesteps, dose_timesteps, init_titres, prop_short, dur_short, dur_long)
  }

  return(titres)
}

ab_r <- function(t, dose_timesteps, init_titres, prop_short, dur_short, dur_long){
  t_index <- findInterval(t, dose_timesteps)
  t_rel <- t - dose_timesteps[t_index]
  rs <- log(2) / dur_short
  rl <- log(2) / dur_long

  ab <- rep(NA, length(t))
  for(i in t){
    t_index_i <- t_index[i]
    t_rel_i <- t_rel[i]
    # Check boost doesn't cause downwards drop in ab titre:
    if(i > 1){
      if(i == dose_timesteps[t_index_i]){
        if(ab[i - 1] > init_titres[t_index_i]){
          init_titres[t_index_i] <- ab[i - 1]
        }
      }
    }
    ab[i] <- init_titres[t_index_i] * ((prop_short[t_index_i] * exp(-rs[t_index_i] * t_rel_i)) + ((1 - prop_short[t_index_i]) *  exp(-rl[t_index_i] * t_rel_i)))
  }

  return(ab)
}
